src/plugins/expressions/common/expression_functions/specs/map_column.ts

```diff
@@ -80,50 +80,50 @@ export const mapColumn: ExpressionFunctionDefinition<
       default: null,
     },
   },
-  fn: (input, args) => {
-    const expression = (...params: Parameters<Required<MapColumnArguments>['expression']>) =>
-      args
-        .expression?.(...params)
-        .pipe(take(1))
-        .toPromise() ?? Promise.resolve(null);
-    const columnId = args.id != null ? args.id : args.name;
+  fn(input, args) {
+    const id = args.id ?? args.name;
 
-    const columns = [...input.columns];
-    const rowPromises = input.rows.map((row) => {
-      return expression({
-        type: 'datatable',
-        columns,
-        rows: [row],
-      }).then((val) => ({
-        ...row,
-        [columnId]: val,
-      }));
-    });
+    return defer(() => {
+      const rows$ = input.rows.length
+        ? zip(
+            ...input.rows.map((row) =>
+              (
+                args.expression?.({
```

Comment:
unrelated to this PR but could maybe still be fixed, expression argument has `required: true` so in type definitions we don't need to make it optional